pkg/cloudprovider/vsphere/loadbalancer/README.md

```diff
@@ -0,0 +1,111 @@
+manager.*
+
+This package enriches the cloud provider interface by implementing the load
+balancing API of the cloud controller for an NSX-T environment.
+
+There are two different modes:
+
+- the *managed* mode manages the load balancer service. The NSX-T load balancer
+  service is only created if it is required.
+  This saves resources if no kubernetes service of type `LoadBalancer` is
+  actually used.
+- the *unmanaged* mode is used if the configuration specifies a load balancer
+  service id. Here only the virtual servers are managed.
+
+The basic assumption is that all nodes are bound to a logical tier1 router.
+Here the load balancer service is attached to. Because there may be only
+one such service here, the configuration of the service must be done
+during the creation of the service. Here the selection of the (t-shirt) size is
+required (S, M or L)  
+For every service port a dedicated virtual server is managed which is
+connected to this load balancer service.
+
+## Features
+
+The load balancer controller part of the vsphere cloud controller manager
+is optional. If no `LoadBalancer` or `LoadBalancerClass` section is given
+in the controller configuration loadbalancing support is disabled.
+If enabled the follwing feature are supported:
+
+### Tagging
+
+All generated NSX-T elements are tagged with the app name of the controller,
+the cluster name and the information from the service. Using this tagging it is
+able to handle recovery of lost or dangling elements and garbage collection of
+unused elements previously generated by the controller, even if the kubernetes
+service object is already (accidentally) gone.
+
+### Load Balancer Classes
+
+This load balancer controller supports the usage of multiple load balancer
+classes. Classes are preconfigured in the configuration file of the cloud
+controller manager. Every class may use another `IPPool` configured in NSX-T.
+This supports the creation of load balancers in different visibility realms,
+for example an `*internet facing* or a *private* load balancer. The
+IPPools must be preconfigured in NSX-T. To select a dedicated loadbalancer
+class the kunernetes service object must be annotated with the annotation:
+
+```yaml
+loadbalancer.vsphere.class: <class name>
+```
+
+If no such annotation is given the default class will be used. The gives
+the adminstrator of the cluster a chance to restrict the usage of the
+NSXT-T resources for cluster users. He can determine which elements should
+be used for a dedicated purpose. The cluster user just needs to know and select
+the purpose by annotating the appropriate load balancer class.
+
+### Health Checks
+
+For TCP load balancers a health check will be generated.
+
+## Configuration File
+
+The controller manager requires dedicated entries in the cloud controller's
+configuration file:
+
+```ini
+[LoadBalancer]
+ipPoolName = pool1
+#ipPoolId = 0815
+lbServiceId = 4711
+size = MEDIUM
+
+[LoadBalancerClass "public"]
```

Comment:
Can we just make this a field in the LoadBalancer config above? So `type=public` or `class=public`?

Reply:
These are arbitrary class names, you can have as many as you want with different settings. The name `public` here is just an example.

Reply:
I see, so where in the config do we point to a specific class. Below I see that `LoadBalancerClassConfig` is embedded in `LoadBalancerConfig` already.

Reply:
The class is selected via annotation on the kubernetes service object of type load balancer.
This is similar to the OpenStack solution. 
The config contains a default, this is part of the flat section. It is used if no class is specified by the service. Then there might be any number of dedicated class specifications in the config file (each in a separate group). The values here are again defaulted by values from the global section. 
So if you just want to use different IPPools and share the same TCP/UDP profiles. The profiles are set in the global section and the IPPools in the dedicated group sections.

Reply:
Okay I think I'm understanding this better now, thanks for explaining :)